doc/SAI-Proposal-Prefix-Compression.md

```diff
@@ -0,0 +1,271 @@
+#  Prefix Compression #
+-------------------------------------------------------------------------------
+ Title       | Prefix Compression
+-------------|-----------------------------------------------------------------
+ Authors     | Nader Shinouda, Cisco
+ Status      | In review
+ Type        | Standards track
+ Created     | 2024-07-1 - Initial Draft
+ SAI-Version | 1.14
+-------------------------------------------------------------------------------
+
+
+## 1.0  Introduction ##
+
+This spec adds Prefix Compression. Prefix Compression allows mapping an IP prefix/mask to a bincode. These prefix/bincode mapping can be grouped together to form prefix compression table.  Tables can have both IPV4 and IPV6 entries.
+
+Prefix Compression tables can be used in features such as ACL to match on a specific bincode. This allows additional functionality to ACL or any feature that can take advantage of such groupings.
+
+## 1.1.0 Function Requirement of Prefix Compression
+- Enable the creation of a prefix compresison table as a SAI object
```

Comment:
SAI infra already has META data assignment in multiple hardware entity like FDB, ROUTE entry etc, Do review them and see if you can add an additional attribute to provide a hint to program the prefix compression table instead. This would minimize the changes in your SAI modelling.

Reply:
I looked at the specs and we should not be touching ROUTE because it will affect the forwarding path. This is unique because it allows the user to create a table based on longest prefix and use that output in other features for lookup. In this case I am using it with ACL but in the future there could be another feature that benefit from this. 

I do agree any time we can minimize SAI changes it would be great, but this is a unique use case and it will add new functionality to SAI. In this case it will add functionality to ACL but again in the future we might find another feature to use this.

Reply:
@rck-innovium, I updated the documentation. I combined both in a single document. Please let me know if this meets your request and resolve if it does

Reply:
I still don't understand the explanation why we can't reuse existing ROUTE Meta data. 

Today one can assign a META data to any route/FDB/.. entry and it won't affect forwarding path. Later this metadata can be used in other features for additional lookup.

Why proposed feature is different?
What functionality SAI will miss if it will continue to be modeled as a META data in ROUTE entry?

Reply:
@ilyav11 

1.  SAI\_ROUTE\_ENTRY\_ATTR\_META\_DATA cannot be assigned based on **Source** IP address. Similarly, FDB metadata cannot be assigned based on the Source MAC address.
2.  For a given packet, a single route entry is matched based on LPM. So, we cannot have separate route entries for forwarding and meta-data.

Reply:
Can you plz elaborate about (2)? 
Do you mean that this table:
- is not used for forwarding, but only to match on pattern that is e.g. SIP/DIP and set meta data - similar to ACL?
- How match in this table done?

Reply:
@ilyav11
- No it is not used for forwarding but a feature that would need to do a lookup based on a LPM. You can do this in ACL today but it will require lots of entries. This allows a customer lets say has a finance pool of addresses, HR pool of addresses and printers pool of addresses. You can use an LPM address for finance, one for HR and one for Printers as an example to represent each department. If you are using ACL then you can easily decide with meta value can access other meta values. Each can be done in a single entry

- This table can be done in HW or SW depending on how the vendor supports. This table would be beneficial for any feature that could take advantage of this LPM mapping to a meta value

We are having a meeting tomorrow to discuss this. Please me know if you want to be included. I can have @tjchadaga  add you to the meeting.